### PR TITLE
feat: add AsyncAPI extension

### PR DESCRIPTION
**Pull request by smoya**
## Description

This PR adds a new extension for [AsyncAPI](https://www.asyncapi.com/). Right now, it only allows the following, but can be easily extended:

- Validate an AsyncAPI Document (from clipboard).
- Validate an AsyncAPI Document (from form input).
- Open an AsyncAPI Document in Studio (from clipboard).
- Open an AsyncAPI Document in Studio (from form input).

## Screencast

https://user-images.githubusercontent.com/1083296/157102752-d2aa6f24-eaea-4c39-81d2-8c405ffc325f.mp4

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed in the `media` folder

### CODE REVIEW

File: extensions/asyncapi/package.json [hunk 1]
```diff
@@ -0,0 +1,59 @@
+{
+  "$schema": "https://www.raycast.com/schemas/extension.json",
+  "name": "AsyncAPI",
+  "title": "AsyncAPI",
+  "description": "AsyncAPI extension for Raycast brings some AsyncAPI tools to you.",
+  "icon": "command-icon.png",
+  "author": "Sergio Moya <smoya89@gmail.com> (sergiomoya.me)",
```
Comment (fe9lix) [inline extensions/asyncapi/package.json:7]: This should just be your Raycast handle / username.

Comment (smoya) [reply extensions/asyncapi/package.json]: Yeah, and it is so clear in the guidelines! Sorry, this is fixed now!

File: extensions/asyncapi/src/validate-clipboard.tsx [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+export default async () => {
+  await showToast(Toast.Style.Animated, "Validating your AsyncAPI document...");
+  const clipboard = await runShellScript("pbpaste");
+  clipboard.length > 0
+    ? await validateAsyncAPIDocument(clipboard)
+    : await showToast(Toast.Style.Failure, "Your clipboard is empty");
+
+  // Commands don’t get their toast handlers called when the command has already been unloaded. This is the case of `no-view` commands,
+  // which they are unloaded right after they have been executed.
+  // In order to let `primaryAction`to work, we need to keep the command running for a bit giving the user time for clicking the action.
```
Comment (pernielsentikaer) [inline extensions/asyncapi/src/validate-clipboard.tsx:14]: This issue will be address in the next release (which is shipped within a day or two, hopefully).

**Review by pernielsentikaer — COMMENTED** [review]
Hi @fearoffish 

Thanks for your contribution 🔥 

I left some feedback and suggestions for you to address 😄 

Let's see when the new version is released, should be very soon!

We should be able to remove the hack, is it okay that we wait until it's released?

**Comment by smoya** [comment]
> Hi @fearoffish
> 
> Thanks for your contribution 🔥
> 
> I left some feedback and suggestions for you to address 😄
> 
> Let's see when the new version is released, should be very soon!
> 
> We should be able to remove the hack, is it okay that we wait until it's released?

Hi @pernielsentikaer, I guess you meant **me** instead of `@fearoffish` 😅 .

I like the idea of waiting for the next version if it finally introduces a way of persisting toasts for some time. Would you mind anyway pointing to the PR with the fix or feature for that so I can read it? Just out of curiosity!

Thanks!

**Comment by pernielsentikaer** [comment]
Haha yeah @smoya, sorry 🙏 

The fix was released in version 1.30.0 a few hours ago:`Toast action handlers 🍞  can now still be called if the toast outlives a dismissed extension` – can you try to remove the hack and check again?

The full [changelog](https://developers.raycast.com/changelog) should be updated tomorrow 😄 

Let me know if you have any questions

**Comment by smoya** [comment]
> Haha yeah @smoya, sorry 🙏
> 
> The fix was released in version 1.30.0 a few hours ago:`Toast action handlers 🍞 can now still be called if the toast outlives a dismissed extension` – can you try to remove the hack and check again?
> 
> The full [changelog](https://developers.raycast.com/changelog) should be updated tomorrow 😄
> 
> Let me know if you have any questions

I removed the hack and now It works! 🎉 🎉  I pushed the changes few secs ago.

Awesome team work folks!

**Comment by pernielsentikaer** [comment]
Hi @smoya 👋 

I'll pass the message to the team 👏

nM2ibzhOocuPmJ15

Here is the full changelog if you want to see, otherwise be sure to join up [Slack](https://raycastcommunity.slack.com/) with us :)

**Comment by smoya** [comment]
Oh, I guess I can now use `Clipboard.readText()` instead of invoking `pbcopy` cmd?

File: extensions/asyncapi/package.json [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      "name": "open-in-studio-clipboard",
+      "title": "Open AsyncAPI Document in Studio",
+      "subtitle": "From clipboard",
+      "description": "Opens the AsyncAPI document from clipboard with the AsyncAPI Studio",
+      "mode": "no-view"
+    }
+  ],
+  "dependencies": {
+    "@asyncapi/parser": "^1.14.1",
+    "@raycast/api": "^1.28.0"
```
Comment (pernielsentikaer) [inline extensions/asyncapi/package.json:39]:
```suggestion
    "@raycast/api": "^1.30.1"
```
Could be cool to use the newest version too 😄

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @smoya 👋 

Awesome to hear that it just works, you didn't update the package, could you do that and see how it goes? 😄

File: extensions/asyncapi/package.json (hunk 3 above)
Comment (smoya) [reply extensions/asyncapi/package.json]: Makes sense as it is required to work 👍

**Comment by smoya** [comment]
> Hi @smoya 👋
> 
> Awesome to hear that it just works, you didn't update the package, could you do that and see how it goes? 😄

Yeah, sorry I forgot to push the changes 😓 . It is done now!

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @smoya 👋 

MUCH better! 🔥 

Since we started working on this extension, we're now adding support for [categories](https://developers.raycast.com/basics/prepare-an-extension-for-store#categories), [screenshots](https://developers.raycast.com/basics/prepare-an-extension-for-store#screenshots) and [changelog](https://developers.raycast.com/basics/prepare-an-extension-for-store#version-history) – do you have time for adding this before proceeding?

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by smoya** [comment]
> Request a new review when you are ready

Hi @pernielsentikaer, I added the screenshots, the `CHANGELOG.md` file, and the categories (in fact just one) into the `package.json` file. 

I hope you can take a look at this :)

**Comment by smoya** [comment]
Hi @pernielsentikaer! is there anything else I should take care of? Thanks!

**Comment by pernielsentikaer** [comment]
Hi @smoya 👋 

If you remove `asyncapi-04.png` then I think we're ready to go 😄 

I'll keep an eye on your PR 💪

**Comment by smoya** [comment]
> Hi @smoya 👋
> 
> If you remove `asyncapi-04.png` then I think we're ready to go 😄
> 
> I'll keep an eye on your PR 💪

What is wrong with that screenshot in particular? I is an screenshot of the tool opened by the extension (AsyncAPI Studio) with the content from clipboard or introduced manually.

I don't think it is a limitation either on the number of screenshots. As per the docs:

> You can add a maximum of six screenshots.

Is the fourth screenshot wrong for any reason?

Thanks!

**Comment by smoya** [comment]
@pernielsentikaer I removed the screenshot file.

As a suggestion, I think you should be more explicit on the requirements. I included that screenshot because of the same particular statement you pasted:

> Select the most informative commands to showcase what your extension does – focus on giving the user as much detail as possible

Thanks for the review!

**Comment by pernielsentikaer** [comment]
Hi @smoya 👋 

I just checked the screenshots again, it seems like they're not in the right sizes with the right amount of padding, did you use the guide under [screenshot](https://developers.raycast.com/basics/prepare-an-extension-for-store#screenshots) to generate them?

I'm sorry, it's my fault, I didn't check it good enough!

I would suggest that you make one new image (of image-02) in the right size to focus on `Select the most informative commands to showcase what your extension does – focus on giving the user as much detail as possible`

If you need help making it, let me know 😄

**Comment by smoya** [comment]
@pernielsentikaer I updated the screenshots `asyncapi-02.png`and `asyncapi-03.png`, however as you can see the `asyncapi-03.png` does not fit the guidelines.
How should we proceed?

**Pull request merged**